### PR TITLE
Hotfix for #422 & # 427 : LLM provider = azureopenai when using detailed report mode

### PR DESCRIPTION
**Pull request by norrisp90**
Hotfix for #422 & #427  Detailed report azureopenai support plus azureopenai LLM settings documentation improvement.

### CODE REVIEW

File: gpt_researcher/llm_provider/azureopenai/azureopenai.py [hunk 1]
```diff
@@ -4,13 +4,18 @@
...
-Please note: Needs additional env vars such as: 
-    AZURE_OPENAI_ENDPOINT  e.g. https://xxxx.openai.azure.com/", 
-    OPENAI_API_VERSION, 
-    OPENAI_API_TYPE
+Please note:
+Needs additional env vars such as: 
+    AZURE_OPENAI_ENDPOINT  e.g. https://xxxx.openai.azure.com/",
+    AZURE_OPENAI_API_KEY e.g "xxxxxxxxxxxxxxxxxxxxx",
+    OPENAI_API_VERSION, e.g. "2024-03-01-preview" but needs to updated over time as API verison updates,
+    AZURE_EMBEDDING_MODEL e.g. "ada2" The Azure OpenAI embedding model deployment name.
```
Comment (Xtrah) [inline gpt_researcher/llm_provider/azureopenai/azureopenai.py:12]:
```suggestion
Please note:
AzureOpenAI requires additional env vars such as: 
    EMBEDDING_PROVIDER="azureopenai"
    LLM_PROVIDER="azureopenai"
    SMART_LLM_MODEL="{LLM_DEPLOYMENT_NAME}"
    FAST_LLM_MODEL="{LLM_DEPLOYMENT_NAME}"
    AZURE_OPENAI_ENDPOINT="https://{ENDPOINT}.openai.azure.com/",
    AZURE_OPENAI_API_KEY="{API_KEY}",
    OPENAI_API_VERSION="2024-02-01",
    AZURE_EMBEDDING_MODEL="{EMBEDDING_MODEL_DEPLOYMENT_NAME}"
    OPENAI_API_TYPE="azure"
```

File: gpt_researcher/llm_provider/azureopenai/azureopenai.py [hunk 2]
```diff
@@ -4,13 +4,18 @@
...
 
-Note new entry in config.py to specify the Azure OpenAI embedding model name:
-self.azure_embedding_model = os.getenv('AZURE_EMBEDDING_MODEL', "INSERT_EMBEDDIGN_MODEL_DEPLOYMENT_NAME")
+config.py settings for Azure OpenAI should look like:
+    self.embedding_provider = os.getenv('EMBEDDING_PROVIDER', 'azureopenai')
+    self.llm_provider = os.getenv('LLM_PROVIDER', "azureopenai")
+    self.fast_llm_model = os.getenv('FAST_LLM_MODEL', "gpt-3.5-turbo-16k") #Deployment name of your GPT3.5T model as per azure OpenAI studio deployment section
+    self.smart_llm_model = os.getenv('SMART_LLM_MODEL', "gpt4")  #Deployment name of your GPT4 1106-Preview+ (GPT4T) model as per azure OpenAI studio deployment section
```
Comment (Xtrah) [inline gpt_researcher/llm_provider/azureopenai/azureopenai.py:18]:
```suggestion
```

**Review by Xtrah — COMMENTED** [review]
Isn't it cleaner to just use a `.env` file considering config.py is made to get environment variables?

Some suggestions below based on that, if you agree.

File: gpt_researcher/llm_provider/azureopenai/azureopenai.py (hunk 2 above)
Comment (norrisp90) [reply gpt_researcher/llm_provider/azureopenai/azureopenai.py:18]: In theory but the config.py pattern is what is currently used and we need to support multiple model types, if you move these Azure OpenAI ones to .env then you are breaking the current design? To implement this properly i think you would have to remove the references to the config.py file values where they are used throughoput the codebase and replace with stright env. var calls?

**Comment by norrisp90** [comment]
> Isn't it cleaner to just use a `.env` file considering config.py is made to get environment variables?
> 
> Some suggestions below based on that, if you agree.

In theory but the config.py pattern is what is currently used, if you move these to .env then you are braking the current design. To implement this properly i think you would have to remove the refences to teh config.py file values where they are used throughoput the codebase and replace with stright env. var calls?

> Isn't it cleaner to just use a `.env` file considering config.py is made to get environment variables?
> 
> Some suggestions below based on that, if you agree.

In theory but the config.py pattern is what is currently used and we need to support multiple model types, if you move these Azure OpenAI ones to .env then you are breaking the current design? To implement this properly i think you would have to remove the references to the config.py file values where they are used throughoput the codebase and replace with stright env. var calls? @Xtrah

**Comment by Xtrah** [comment]
> > Isn't it cleaner to just use a `.env` file considering config.py is made to get environment variables?
> > Some suggestions below based on that, if you agree.
> 
> In theory but the config.py pattern is what is currently used, if you move these to .env then you are breakinh the current design. To implement this properly i think you would have to remove the references to the config.py file values where they are used throughoput the codebase and replace with stright env. var calls? @Xtrah

Isn't it supported by the current design, considering you are using:

```
self.smart_llm_model = os.getenv('SMART_LLM_MODEL', "gpt-4-1106-preview")
```

Where the optional second argument is the default. So you can either specify changes in .env or you can change the default, both with the same result?

**Comment by norrisp90** [comment]
> > > Isn't it cleaner to just use a `.env` file considering config.py is made to get environment variables?
> > > Some suggestions below based on that, if you agree.
> > 
> > 
> > In theory but the config.py pattern is what is currently used, if you move these to .env then you are breakinh the current design. To implement this properly i think you would have to remove the references to the config.py file values where they are used throughoput the codebase and replace with stright env. var calls? @Xtrah
> 
> Isn't it supported by the current design, considering you are using:
> 
> ```
> self.smart_llm_model = os.getenv('SMART_LLM_MODEL', "gpt-4-1106-preview")
> ```
> 
> Where the `os.getenv` optional second argument is the default. So you can either specify changes in .env or you can change the default, both with the same result? That's why I was thinking it is cleaner to just do changes in one file that is already dynamic, rather than the code.

Not really. If you place something in env var it will only work if it's called directly as env in the code. Calls like config.llm_provider are used extensively also which would all need to be replaced if everything was to move to env variables.

Your idea on defaults also doesn't consider when llm_provider type is set to another model provider other than Azure OpenAI which work differently and would need other defaults or none at all etc.

**Review by deb1305 — COMMENTED** [review]
this fix is working

**Review by Xtrah — APPROVED** [review]
@norrisp90 considering config.py uses os.getenv for all variables, and we are only calling on config.variable in the code, it will work both by setting new defaults in config.py and setting a variable in .env. Setting the variable in .env will overwrite whatever is set in config.py. An advantage of using .env allows to e.g. host a Docker image on GHCR or Docker Hub that is configurable right out of the box with just .env and no code changes, opening up for more users.

Thank you for fixing this so quickly, I have confirmed [it works great](https://gist.github.com/Xtrah/167dbd3ab0eb5f38fa39006dc27d2c12)!

**Comment by assafelovic** [comment]
Thank you @norrisp90 !

**Pull request merged**